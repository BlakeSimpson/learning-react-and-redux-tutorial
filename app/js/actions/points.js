import { FAVOURITE_ADDED, FAVOURITE_REMOVED } from '../constants';

export const addFavourite = id => {
  return {
    type: FAVOURITE_ADDED,
    payload: {
      id: id
    }
  };
};

export const removeFavourite = id => {
  return {
    type: FAVOURITE_REMOVED,
    payload: {
      id: id
    }
  };
};
