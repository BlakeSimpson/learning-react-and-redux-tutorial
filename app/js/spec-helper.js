import 'raf/polyfill';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

export const pointsMock = [
  {
    id: 1,
    x: 450,
    y: 110,
    details: {
      name: 'The Wall',
      house: "Night's Watch",
      words: 'Night gathers, and now my watch begins.'
    },
    favourite: true
  },
  {
    id: 2,
    x: 375,
    y: 355,
    details: {
      name: 'Winterfell',
      house: 'Stark',
      words: 'Winter is Coming'
    },
    favourite: false
  }
];
