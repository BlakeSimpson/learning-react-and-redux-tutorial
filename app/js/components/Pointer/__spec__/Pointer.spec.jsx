import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';

import { pointsMock } from '../../../spec-helper';
import { addFavourite, removeFavourite } from '../../../actions';
import ConnectedPointer, { Pointer } from '../';

const noop = () => {};

const pointerProps = {
  addFavourite: noop,
  removeFavourite: noop,
  id: 42,
  x: 99,
  y: 88,
  details: {
    name: 'Somewhere',
    house: 'Bungalow',
    words: 'Hello'
  },
  favourite: true
};

describe('Pointer', () => {
  it('matches the snapshot', () => {
    const wrapper = shallow(<Pointer {...pointerProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('when the toggle button is clicked', () => {
    it('changes open state to true, if it is false', () => {
      const wrapper = mount(<Pointer {...pointerProps} />);

      wrapper.find('a[href="#toggle"]').simulate('click');

      expect(wrapper.state('open')).toEqual(true);
    });

    it('changes open state to true, if it is false', () => {
      const wrapper = mount(<Pointer {...pointerProps} />);

      wrapper.setState({ open: true });
      wrapper.find('a[href="#toggle"]').simulate('click');

      expect(wrapper.state('open')).toEqual(false);
    });
  });
});

describe('ConnectedApp', () => {
  const mockStore = configureStore([]);
  const store = mockStore({ points: pointsMock });

  beforeEach(() => {
    store.clearActions();
  });

  it('maps store state to the props', () => {
    const wrapper = shallow(<ConnectedPointer store={store} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('when the favourite button is clicked', () => {
    it('calls the removeFavourite action if the pointer is a favourite', () => {
      const wrapper = mount(
        <ConnectedPointer {...pointerProps} store={store} />
      );
      const expectedAction = [removeFavourite(42)];

      wrapper.find('a[href="#favourite"]').simulate('click');

      expect(store.getActions()).toEqual(expectedAction);
    });

    it('calls the removeFavourite action if the pointer is a favourite', () => {
      const abc = { favourite: false };
      const modifiedPointerProps = Object.assign({}, pointerProps, abc);

      const wrapper = mount(
        <ConnectedPointer {...modifiedPointerProps} store={store} />
      );
      const expectedAction = [addFavourite(42)];

      wrapper.find('a[href="#favourite"]').simulate('click');

      expect(store.getActions()).toEqual(expectedAction);
    });
  });
});
