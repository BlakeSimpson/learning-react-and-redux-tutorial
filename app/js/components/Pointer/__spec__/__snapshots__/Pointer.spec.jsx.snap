// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectedApp maps store state to the props 1`] = `
<ConnectedPointer
  addFavourite={[Function]}
  removeFavourite={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;

exports[`Pointer matches the snapshot 1`] = `
<div
  className="undefined"
  onClick={[Function]}
  style={
    Object {
      "left": 99,
      "top": 88,
    }
  }
>
  <div
    className="undefined"
  >
    <header>
      <h3>
        Somewhere
      </h3>
      <div>
        <a
          href="#favourite"
          onClick={[Function]}
        >
          –
        </a>
        <a
          href="#toggle"
          onClick={[Function]}
        >
          ×
        </a>
      </div>
    </header>
    <p>
      House: 
      Bungalow
    </p>
    <p>
      Words: 
      Hello
    </p>
  </div>
</div>
`;
